### PR TITLE
[for fleet provisioning with CSR to support write key and certificate to disk] support optional write generated private key

### PR DESCRIPTION
**Pull request by giuspen**
Support optional write generated private key to disk

Description
-----------
For use by the fleet provisioning with CSR demo ( see https://github.com/giuspen/aws-iot-device-sdk-embedded-C/tree/GP_fleet_provisioning_with_csr_demo/demos/fleet_provisioning/fleet_provisioning_with_csr ), I would like to support writing key and certificate to disk when the optional macros DOWNLOADED_CERT_WRITE_PATH and GENERATED_PRIVATE_KEY_WRITE_PATH are provided

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [ ] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

### CODE REVIEW

File: source/portable/mbedtls/core_pkcs11_mbedtls.c [hunk 1]
```diff
@@ -55,6 +55,11 @@
 /* C runtime includes. */
 #include <string.h>
 
+#if defined( GENERATED_PRIVATE_KEY_WRITE_PATH )
+    #include <errno.h>
+    #define PRIV_KEY_BUFFER_LENGTH                         2048
+#endif // GENERATED_PRIVATE_KEY_WRITE_PATH
```
Comment (Skptak) [inline source/portable/mbedtls/core_pkcs11_mbedtls.c:61]:
```suggestion
#endif /* GENERATED_PRIVATE_KEY_WRITE_PATH */
```

File: source/portable/mbedtls/core_pkcs11_mbedtls.c [hunk 2]
```diff
@@ -5676,6 +5681,40 @@ CK_DECLARE_FUNCTION( CK_RV, C_GenerateKeyPair )( CK_SESSION_HANDLE hSession,
...
+                size_t privatekeyLength = strlen( privatekey );
+                FILE* fp = fopen( GENERATED_PRIVATE_KEY_WRITE_PATH, "w" );
+
+                if( NULL != fp )
+                {
+                    const size_t writtenBytes = fwrite( privatekey, 1u, privatekeyLength, fp );
+
+                    if( writtenBytes == privatekeyLength )
+                    {
+                        LogInfo( ( "Written %s successfully.", GENERATED_PRIVATE_KEY_WRITE_PATH ) );
```
Comment (Skptak) [inline source/portable/mbedtls/core_pkcs11_mbedtls.c:5700]:
```suggestion
                        LogInfo( ( "Wrote the generate private key to %s successfully.", GENERATED_PRIVATE_KEY_WRITE_PATH ) );
```

**Comment by Skptak** [comment]
Hey, thanks for submitting all the PRs you've raised to add in this feature!
I've commented two smaller nits that I saw in your changes on this PR.

I do like this idea, as I can see how useful it is for debugging. That said, I think using this feature in the file should generate a `#warning` just to make sure that it isn't accidentally left in.

**Comment by giuspen** [comment]
Thanks @Skptak for the review I'll also add the compilation warning shortly

**Review by manvensh — APPROVED** [review]

**Review by Skptak — APPROVED** [review]

**Pull request merged**